src/base/bios/bios.S

```diff
@@ -165,6 +165,10 @@ PKTDRV_stats:
 LFN_short_name:
 	.space 128
 
+	.globl	LFN_long_name
+LFN_long_name:
+	.space 128
```

Comment:
It probably requires more space - IIRC
this is the name with full path (please check).

Reply:
But I meant to put here only the
pointer to LFN. The LFN itself is passed
from the caller, so why to copy it, wasting
precious bios memory space?

Reply:
I didn't check for a short fix this time, but I did fix it before in the original non applied PR #485 and I think file permissions are wrong. In fact I think use of get_unix_attr() is bogus even now because I open a file for write and I get perms 0444.

The filename fields in the SDA are 128 bytes long. Since the redirector is expected to use those to work with, I can't see how DOS would allow anything longer.

The caller does not have the fully qualified pathname, only what the user asked for. So not necessarily any path or drive (it could be relative). That said, you could just save the filename itself and reassemble it from the short filename.

Reply:
But redirector is never expected to work
with LFNs, to start with. Please see here:
https://en.wikipedia.org/wiki/Long_filename
```
The long filename system allows a maximum length of 255 UCS-2 characters
...
The maximum length of a pathname is 256 characters, which includes all parent directories and the filename.
```
So I guess the buffer should be 512
bytes, but why to do the buffering at all
if we can store pointer?

> That said, you could just save the filename itself and reassemble it from the short filename.

But why would I do that?
Besides, reassembling the path that contains
the mangled dir names, is difficult. Yes, I know
the code that does it, is already there, but for what?

Reply:
> But redirector is never expected to work with LFNs

Like it or not that's the interface we have. I just don't see how DOS could pass us a pathname over SDA > 128 byte, unless there's a new version of SDA for Win95+? If you are concerned about buffer overrun, then checking the length before storing is a good idea.

> But why would I do that?

For example if you are concerned about space then as longname and shortname share common drive and path, just pass the full shortname as we do now and pass just the difference in another smaller var.
~~~
LFN: open longname = 'D:\verylongname.txt'                                      
LFN: open shrtname = 'D:\VERYL~D#.TXT'
~~~

Bear in mind that the caller never passed 'D:\verylongname.txt' in the open call, only 'verylongfilename.txt', dosemu figured out the path and drive. So I don't think you can just pass a pointer via bios.S I think you need to allocate some DOS memory. Or am I wrong?

Reply:
> Like it or not that's the interface we have.

What kind of? LFN interface has no connection to
the redirector interface, and if you don't trust wikipedia
quotes, here is the RBIL quote:
```
INT 21 - Windows95 - LONG FILENAME - GET VOLUME INFORMATION
        AX = 71A0h
        DS:DX -> ASCIZ root name (e.g. "C:\")
        ES:DI -> buffer for file system name
        CX = size of ES:DI buffer
Return: CF clear if successful
            AX destroyed (0000h and 0200h seen)
            BX = file system flags (see #01783)
            CX = maximum length of file name [usually 255]
            DX = maximum length of path [usually 260]
```

> I just don't see how DOS could pass us a pathname over SDA > 128 byte,

Please see the 0x71 API, it passes ASCIIZ via
DS:SI pointers. Its simple.

> unless there's a new version of SDA for Win95+? 

No and I don't see why it should.

> 'verylongfilename.txt', dosemu figured out the path and drive.

Even that is 255. And if it is UCS2, then likely
multiple by 2 is needed.

> So I don't think you can just pass a pointer via bios.S
> I think you need to allocate some DOS memory. Or am I wrong?

Please name the exact 0x71 function that
does not pass LFN via pointer that can be
easily stored anywhere. I don't see what you
mean.

Reply:
cd \very\long\directory
run program containing open("verylongname.txt")
~~~
LFN: open 'verylongname.txt'  # note no drive or path on int21/716c
LFN: doing LFN!, AX=716c DL=1                                                   
LFN: open 'verylongname.txt'                                                    
LFN: build_truename: verylongname.txt                                           
truename(verylongname.txt)                                                      
CDS entry: #3 @0xb233dd48 (2) 'D:\VERY\LONG\DIREC~A5'                           
MFS: is_msdos_device verylong                                                   
SUBSTing from: D:\VERY\LONG\DIREC~A5                                            
Absolute logical path: "D:\VERY\LONG\DIREC~A5\verylongname.txt"
~~~

So what I see from this is
- I now agree that SDA length of 128 is probably okay because the full path will be mangled into something shorter, hence a redirector could potentially unmangle back to upto whatever limit is on its local filesystem.
- I still don't agree that int21/716c Open has the full path string to the file, yes it has an LFN but it may not have drive and current directory prepended to it. Perhaps it's okay to figure that out in the redirector? If so your pointer passing should be okay.
- Regarding UCS2 - how does int21/71xx know if you passed it UCS2 chars or a sequence of 8bit chars?

Reply:
Andrew, I never said it needs full path!
Please, if you see the above RBIL quote,
you'll find that even the name itself is up
to 255 chars!
I think UCS2 is used only to store the
names on the real (vfat) FS, but not passed
to DOS API, and even then its 255 bytes.

Reply:
255 chars + '\0' - 256 bytes minimum.
See?

Reply:
OK, not minimum but maximum.
Doesn't matter. :)

Reply:
> I now agree that SDA length of 128 is probably okay because the full path
> will be mangled into something shorter, hence a redirector could potentially
> unmangle back to upto whatever limit is on its local filesystem.

Unmangle back needs LFN file already
created. And my suggestion is exactly to
avoid creating LFN file beforehand, so I
don't think "unmangle back" is a good
idea, even if the name doesn't include
the full path.

Reply:
Yes, sure I'm with you now. I wasn't suggesting that we do unmangle, just to notice that the SDA contains a mangled path, not a plain one, and that it might represent a path longer than 128 bytes long.
BTW should the bios.S LFN long name pointer store a dosaddr_t or be seg and ofs?

Reply:
> SDA contains a mangled path

In our case yes, because our helper
does so. In plain DOS, as Bart explained
recently, there is no connection between
LFN and mfs at all. To me that means
that if you map the network drive under
windows, then DOS progs will see only
SFNs on it. Unless, maybe, winoldap
provides something useful?
I think freedos played with an idea of
integrating LFNs, but its not there.

> BTW should the bios.S LFN long name pointer store a dosaddr_t or be seg and ofs?

Of course it should be seg:ofs because
client passes it in a seg:ofs format. There
is no need to do any extra work.
I suggest this to look up the original SFN
pointer:
http://www.ctyme.com/intr/rb-4392.htm